src/MIDebugEngine/Engine.Impl/Disassembly.cs

```diff
@@ -305,17 +305,145 @@ private void DeleteRangeFromCache(DisassemblyBlock block)
             }
         }
 
+        private class DisasmAddressRange
+        {
+            public readonly string Symbol;
+            public ulong StartAddress;
+            public ulong EndAddress;
+            private readonly Dictionary<ulong, DisasmInstruction> _AddressToInstruction;
+
+            public DisasmAddressRange(DisasmInstruction instruction)
+            {
+                Symbol = instruction.Symbol;
+                StartAddress = instruction.Addr;
+                EndAddress = instruction.Addr + 1;
+                _AddressToInstruction = new Dictionary<ulong, DisasmInstruction>()
+                {
+                    {instruction.Addr, instruction}
+                };
+            }
+
+            public void UpdateEndAddress(DisasmInstruction instruction)
+            {
+                EndAddress = instruction.Addr + 1;
+                _AddressToInstruction.Add(instruction.Addr, instruction);
+            }
+
+            public void MapSourceToInstructions(DebuggedProcess process, TupleValue[] src_and_asm_lines)
+            {
+                /* Example response
+                 *  [
+                 *   {
+                 *      line="15",
+                 *      file="main.cpp",
+                 *      fullname="/home/cpp/main.cpp",
+                 *      line_asm_insn=[
+                 *          {
+                 *              address="0x0000000008001485",
+                 *              func-name="main(int, char**)",
+                 *              offset="316",
+                 *              opcodes="83 bd 4c ff ff ff 00",
+                 *              inst="cmpl   $0x0,-0xb4(%rbp)"
+                 *          },
+                 *          {
+                 *              address="0x000000000800148c",
+                 *              func-name="main(int, char**)",
+                 *              offset="323",
+                 *              opcodes="75 07",
+                 *              inst="jne    0x8001495 <main(int, char**)+332>"
+                 *          }
+                 *      ]
+                 *   }
+                 *  ]
+                 */
+                foreach (TupleValue src_and_asm_line in src_and_asm_lines)
+                {
+                    uint line = src_and_asm_line.FindUint("line");
```

Comment:
Watch out, this can fail (Exception) because "line" can be missing. When "line" is missing so will the "file" or "filename". I did it backwards. First I check for the file and if it exists and the line exists, then I call `FindUnit("line")`. It happens when a few instructions have source info but a few don't in the same addr-rane. If all don't have source-info, then there will be no  `src_and_asm_line` either. [My code looks like](https://github.com/haneefdm/MIEngine/blob/507a8140aa0de4f02e20fd10387da084d5fddc9e/src/MIDebugEngine/Engine.Impl/Disassembly.cs#L343).

```C#
                    string file = process.GetMappedFileFromTuple(item);
                    uint line = !String.IsNullOrEmpty(file) && item.Contains("line") ? item.FindUint("line") : 0;
```

Reply:
Oh, I see you are disassembling twice. I did not find that necessary, but you may have seen things I haven't. You can ignore my previous comment but I would still guard against not having file/source info in the middle of normal looking stuff.